src/Microsoft.DotNet.Build.Tasks/RemoveDuplicatesWithLastOneWinsPolicy.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    /// <summary>
+    /// Given a list of items, remove duplicate items. Metadata is not considered. Case insensitive.
+    /// Unlike <see cref="RemoveDuplicates"/>, the last item in the list wins rather than the first.
+    /// That distinction is important when the items have different metadata.
+    /// </summary>
+    public class RemoveDuplicatesWithLastOneWinsPolicy : Task
+    {
+        /// <summary>
+        /// The list of items from which to remove duplicates.
+        /// </summary>
+        public ITaskItem[] Inputs
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The list of items with duplicates removed.
+        /// </summary>
+        [Output]
+        public ITaskItem[] Filtered
+        {
+            get;
+            set;
+        }
+
+        public override bool Execute()
+        {
+            var existingIndexMap = new Dictionary<string, int>(Inputs.Length, StringComparer.OrdinalIgnoreCase);
+            var filteredList = new List<ITaskItem>(Inputs.Length);
+
+            foreach (ITaskItem item in Inputs)
+            {
+                int existingIndex;
+                if (existingIndexMap.TryGetValue(item.ItemSpec, out existingIndex))
+                {
+                    filteredList[existingIndex] = item;
+                }
+                else
+                {
+                    filteredList.Add(item);
+                    existingIndexMap.Add(item.ItemSpec, filteredList.Count - 1);
+                }
+            }
+
+            Filtered = filteredList.ToArray();
+            return true;
```

Comment:
As a complete aside, the proposal https://github.com/dotnet/corefx/issues/2146 would help here.  With a DistinctBy method, this whole thing could become:
```C#
filtered = Inputs.DistinctBy(item => item.ItemSpec, StringComparer.OrdinalIgnoreCase).ToArray();
return true;
```
It could also be simplified if we had the equivalent of ```Comparer<T>.CreateComparer``` but for ```IEqualityComparer<T>```, e.g. ```public static IEqualityComparer<T> Create(Func<T, int> getHashCode, Func<T, T, bool> equals);```.  With something like that, you could do the same thing with Distinct.